File: backend/app/schemas/resume.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field

class ContactInfo(BaseModel):
    email: Optional[str] = None
    phone: Optional[str] = None
    linkedin: Optional[str] = None
    github: Optional[str] = None
    portfolio: Optional[str] = None
    location: Optional[str] = None

class EducationItem(BaseModel):
    institution: str
    degree: Optional[str] = None
    field_of_study: Optional[str] = None
    graduation_year: Optional[str] = None
    gpa: Optional[str] = None

class WorkExperienceItem(BaseModel):
    company: str
    role: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    duration: Optional[str] = None
    achievements: List[str] = Field(default_factory=list)

class ProjectItem(BaseModel):
    name: str
    description: Optional[str] = None
    technologies: List[str] = Field(default_factory=list)
    link: Optional[str] = None

class KeywordAnalysis(BaseModel):
    top_keywords: List[str] = Field(default_factory=list)
    density: Dict[str, float] = Field(default_factory=dict)
    industry_terms: List[str] = Field(default_factory=list)

class ResumeAnalysisBase(BaseModel):
    candidate_name: Optional[str] = None
    contact_info: ContactInfo = Field(default_factory=ContactInfo)
    summary: Optional[str] = None
    education: List[EducationItem] = Field(default_factory=list)
    work_experience: List[WorkExperienceItem] = Field(default_factory=list)
    projects: List[ProjectItem] = Field(default_factory=list)
    certifications: List[str] = Field(default_factory=list)
    technical_skills: List[str] = Field(default_factory=list)
    soft_skills: List[str] = Field(default_factory=list)
    tools_and_technologies: List[str] = Field(default_factory=list)
    
    resume_score: int = 0
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)
    improvement_suggestions: List[str] = Field(default_factory=list)
    missing_sections: List[str] = Field(default_factory=list)
    keyword_analysis: KeywordAnalysis = Field(default_factory=KeywordAnalysis)

class ResumeAnalysisOut(ResumeAnalysisBase):
    id: str
    resume_id: str
    created_at: datetime

    class Config:
        from_attributes = True

class ResumeBase(BaseModel):
    file_name: str
    file_type: str
    file_size_bytes: int

class ResumeOut(ResumeBase):
    id: str
    raw_text: str
    created_at: datetime
    updated_at: datetime
    analysis: Optional[ResumeAnalysisOut] = None

    class Config:
        from_attributes = True

class ResumeListItem(ResumeBase):
    id: str
    candidate_name: Optional[str] = None
    resume_score: Optional[int] = None
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/resume.ts
export interface ContactInfo {
  email?: string | null;
  phone?: string | null;
  linkedin?: string | null;
  github?: string | null;
  portfolio?: string | null;
  location?: string | null;
}

export interface EducationItem {
  institution: string;
  degree?: string | null;
  field_of_study?: string | null;
  graduation_year?: string | null;
  gpa?: string | null;
}

export interface WorkExperienceItem {
  company: string;
  role: string;
  start_date?: string | null;
  end_date?: string | null;
  duration?: string | null;
  achievements: string[];
}

export interface ProjectItem {
  name: string;
  description?: string | null;
  technologies: string[];
  link?: string | null;
}

export interface KeywordAnalysis {
  top_keywords: string[];
  density: Record<string, number>;
  industry_terms: string[];
}

export interface ResumeAnalysis {
  id: string;
  resume_id: string;
  candidate_name?: string | null;
  contact_info: ContactInfo;
  summary?: string | null;
  education: EducationItem[];
  work_experience: WorkExperienceItem[];
  projects: ProjectItem[];
  certifications: string[];
  technical_skills: string[];
  soft_skills: string[];
  tools_and_technologies: string[];
  resume_score: number;
  strengths: string[];
  weaknesses: string[];
  improvement_suggestions: string[];
  missing_sections: string[];
  keyword_analysis: KeywordAnalysis;
  created_at: string;
}

export interface Resume {
  id: string;
  file_name: string;
  file_type: string;
  file_size_bytes: number;
  raw_text: string;
  created_at: string;
  updated_at: string;
  analysis?: ResumeAnalysis | null;
}

export interface ResumeListItem {
  id: string;
  file_name: string;
  file_type: string;
  file_size_bytes: number;
  candidate_name?: string | null;
  resume_score?: number | null;
  created_at: string;
}
